### PR TITLE
feat: add chunk row accessors for typed and string reads

### PR DESCRIPTION
**Pull request by mrchypark**
## Summary
- add `ChunkRows` plus `Conn.QueryChunksContext` and `Stmt.QueryChunksContext` for chunk-by-chunk result access
- add typed `DataChunk` accessors for `INTEGER`, `BIGINT`, `DOUBLE`, `TIMESTAMP`, and zero-copy `VARCHAR` reads
- add tests and benchmarks covering chunk iteration, null handling, type mismatches, and chunk-oriented performance

## Motivation
The existing `database/sql` row path is still the right default, but some consumers need a lower-overhead read path when they already process results chunk-wise. This change keeps the existing API untouched and adds an opt-in chunk API on top of the existing `DataChunk` abstraction.

The typed accessors are intentionally narrow:
- `IsNull`
- `Int32Slice`
- `Int64Slice`
- `Float64Slice`
- `TimestampMicrosSlice`
- `StringRefs`

All returned views are documented as valid only until the next `NextChunk` call or `Close`.

## Benchmarks
On an Apple M1 with:

```sh
go test ./... -run '^$' -bench 'BenchmarkQueryChunksContext(TypedNumericScan|GetValueNumericScan|TypedStringScan|Count)$' -benchmem -benchtime=1x
```

- `BenchmarkQueryChunksContextGetValueNumericScan`: `37.57 ms/op`, `1.06 MB/op`, `131162 allocs/op`
- `BenchmarkQueryChunksContextTypedNumericScan`: `4.66 ms/op`, `13.9 KB/op`, `346 allocs/op`
- `BenchmarkQueryChunksContextTypedStringScan`: `13.51 ms/op`, `1.07 MB/op`, `444 allocs/op`
- `BenchmarkQueryChunksContextCount`: `4.71 ms/op`, `18.1 KB/op`, `411 allocs/op`

The main improvement is removing per-cell boxing and `GetValue` overhead when the caller can already consume chunk-oriented data.

## Validation
- `go test ./...`
- `go test ./... -run '^$' -bench 'BenchmarkQueryChunksContext(TypedNumericScan|GetValueNumericScan|TypedStringScan|Count)$' -benchmem -benchtime=1x`

### CODE REVIEW

**Comment by cypark-conalog** [comment]
I opened this as a draft because I wanted to get early feedback on the interface shape before treating it as a finished proposal.

If you think this should be smaller, split differently, or exposed through a different API, I would be happy to follow that direction. In particular, if there is a preferred way to introduce the chunk access path or the typed accessors, please let me know and I can adjust the PR accordingly.

I would appreciate any guidance on how you would like this to be shaped before a full review starts.

**Comment by mlafeldt** [comment]
> I would appreciate any guidance on how you would like this to be shaped before a full review starts.

Thanks for the draft. I think the overall shape looks good. Keeping it as an explicit low-level, opt-in path is the right call. Please keep it focused. We can add more typed accessors later. Maybe add a test for empty and non-inlined strings as well.

**Comment by mrchypark** [comment]
> > I would appreciate any guidance on how you would like this to be shaped before a full review starts.
> 
> Thanks for the draft. I think the overall shape looks good. Keeping it as an explicit low-level, opt-in path is the right call. Please keep it focused. We can add more typed accessors later. Maybe add a test for empty and non-inlined strings as well.

Thanks, that makes sense. I'll keep this PR focused and add coverage for empty strings and non-inlined strings in the StringRefs tests.

File: data_chunk_typed.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+// Float64Slice returns a zero-copy view over a DOUBLE column.
+// The returned slice is only valid until the chunk is invalidated.
+// For NULL rows, use IsNull to check whether the corresponding element is valid.
+func (chunk *DataChunk) Float64Slice(colIdx int) ([]float64, error) {
+	return typedChunkSlice[float64](chunk, colIdx, TYPE_DOUBLE)
+}
+
+// TimestampMicrosSlice returns TIMESTAMP values as Unix microseconds.
+// The returned slice is only valid until the chunk is invalidated.
```
Comment (mlafeldt) [inline data_chunk_typed.go]: This comment is wrong about ownership. The slice is allocated via `make`.

File: data_chunk_typed.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
+package duckdb
+
+import (
+	"unsafe"
+
+	"github.com/duckdb/duckdb-go/v2/mapping"
+)
+
+// IsNull reports whether the value at colIdx/rowIdx is NULL.
+func (chunk *DataChunk) IsNull(colIdx, rowIdx int) (bool, error) {
```
Comment (mlafeldt) [inline data_chunk_typed.go:10]: This function should also validate `rowIdx` to avoid out-of-range row indexes.

Reject `rowIdx < 0 || rowIdx >= chunk.GetSize()`

File: data_chunk_string.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+func (chunk *DataChunk) StringRefs(colIdx int) ([]StringRef, error) {
+	values, err := typedChunkSlice[mapping.StringT](chunk, colIdx, TYPE_VARCHAR)
+	if err != nil {
+		return nil, err
+	}
+
+	refs := make([]StringRef, len(values))
+	for idx := range values {
+		refs[idx] = newStringRef(&values[idx])
+	}
```
Comment (mlafeldt) [inline data_chunk_string.go:57]: This currently crashes randomly, see `go test -run TestQueryChunksContextStringRefs -count=20`.

Fix:

```suggestion
	for idx := range values {
		if chunk.columns[colIdx].getNull(mapping.IdxT(idx)) {
			continue
		}
		refs[idx] = newStringRef(&values[idx])
	}
```

**Review by mlafeldt — CHANGES_REQUESTED** [review]
Thanks for this. A few things to fix before merge:

1. `StringRefs` crashes on NULL VARCHAR rows
2. `IsNull` doesn't validate `rowIdx`
3. `TimestampMicrosSlice` doc is inaccurate
4. CI fails: `golangci-lint run` reports 8 issues

See my inline comments.

File: data_chunk_string.go (hunk 3 above)
Comment (mrchypark) [reply data_chunk_string.go:57]: Applied in b93f4bf. StringRefs now skips NULL rows before constructing StringRef values, so NULL VARCHAR rows keep the zero-value ref. Verified with go test -run 'TestQueryChunksContext(IsNullRejectsOutOfRangeRowIndex|StringRefs)$' -count=20, go test ./..., golangci-lint run, and the local static-lib test command.

File: data_chunk_typed.go (hunk 2 above)
Comment (mrchypark) [reply data_chunk_typed.go:10]: Applied in b93f4bf. IsNull now rejects rowIdx < 0 and rowIdx >= chunk.GetSize() before converting to mapping.IdxT, with regression coverage for both bounds. Verified with go test -run 'TestQueryChunksContext(IsNullRejectsOutOfRangeRowIndex|StringRefs)$' -count=20, go test ./..., and golangci-lint run.

File: data_chunk_typed.go (hunk 1 above)
Comment (mrchypark) [reply data_chunk_typed.go]: Applied in 25a2419. Updated the TimestampMicrosSlice documentation to state that it returns a copy, matching the make-backed implementation. Verified with go test -run TestQueryChunksContextTimestampMicrosSlice, go test ./..., and golangci-lint run.

**Comment by hiidy** [comment]
Thank you for the careful review. We really need this chunk API for our use case.

The current row-based approach doesn't work for our workloads, and this would be a game-changer 
for us. All the feedback has been addressed, and we'd be incredibly grateful if this could 
be merged.

@mlafeldt  thanks so much for considering!
